File: VelibStations_iOS/Common/View/BaseViewModel.swift
//
//  BaseViewModel.swift
//  VelibStations
//
//  Created by Safi on 15/10/2019.
//

import Foundation
import UIKit
import CommonKit

class BaseViewModel : IBaseViewModel {
    
    private var mModel: IBaseModel? = nil
    private var mEventSubscriber: IEventSubscriber? = nil
    
    required init() {
        
    }
    
    func getModel() -> IBaseModel? {
        mModel
    }
    
    func setModel(model: String) {
        mModel = (NSClassFromString(model) as! BaseModel.Type).init()
        mModel?.onLoad()
    }
    
    func onLoad() {
        // Essential addons for activity model.
        addAddOn(type: AddOnType().EVENT_MANAGER, addOn: AddOnManager().getAddOn(type: AddOnType().EVENT_MANAGER)!)
    }
    
    func onUnload() {
        // Unload model.
        mModel?.onUnload()
        
        clearAddOns()
        receiveEvents(receive: false)
    }
    
    func receiveEvents(receive: Bool) {
        let eventManager: IEventManager? = getAddOn(type: AddOnType().EVENT_MANAGER) as? IEventManager
        
        if (receive) {
            mEventSubscriber = eventManager?.subscribe(callback: { event in
                self.onReceiveEvents(event: event)
            })
        } else {
            if (mEventSubscriber != nil) {
                eventManager?.unsubscribe(eventSubscriber: mEventSubscriber)
            }
            mEventSubscriber = nil
        }
    }
    
    func onReceiveEvents(event: Event) {
        
    }
    
    // Addons related methods.
    private let mAddOn: IAddOn = AddOn()
    
    func getAddOn(type: String) -> IAddOn? {
        return mAddOn.getAddOn(type: type)
    }
    
    func getAddOns() -> [String : IAddOn] {
        return mAddOn.getAddOns()
    }
    
    func addAddOn(type: String, addOn: IAddOn) {
        mAddOn.addAddOn(type: type, addOn: addOn)
    }
    
    func addAddOns(addons: [String : IAddOn]) {
        mAddOn.addAddOns(addons: addons)
    }
    
    func removeAddOn(type: String) {
        mAddOn.removeAddOn(type: type)
    }
    
    func removeAddOns(types: [String]) {
        mAddOn.removeAddOns(types: types)
    }
    
    func clearAddOns() {
        mAddOn.clearAddOns()
    }
}

File: VelibStations_Android/app/src/main/java/com/muhammedsafiulazam/common/view/BaseViewModel.kt
package com.muhammedsafiulazam.common.view

import androidx.lifecycle.ViewModel
import com.muhammedsafiulazam.common.addon.AddOn
import com.muhammedsafiulazam.common.addon.AddOnManager
import com.muhammedsafiulazam.common.addon.AddOnType
import com.muhammedsafiulazam.common.addon.IAddOn
import com.muhammedsafiulazam.common.event.Event
import com.muhammedsafiulazam.common.event.IEventManager
import com.muhammedsafiulazam.common.event.IEventSubscriber

/**
 * Created by Muhammed Safiul Azam on 24/07/2019.
 */

open class BaseViewModel : ViewModel(), IBaseViewModel {
    private var mModel: IBaseModel? = null
    private var mEventSubscriber: IEventSubscriber? = null


    override fun getModel(): IBaseModel? {
        return mModel
    }

    override fun setModel(model: String) {
        mModel = Class.forName(model).newInstance() as IBaseModel?
        mModel?.onLoad()
    }

    override fun onLoad() {
        // Essential addons.
        addAddOn(AddOnType.EVENT_MANAGER, AddOnManager.getAddOn(AddOnType.EVENT_MANAGER)!!)
    }

    override fun onUnload() {
        // Unload model.
        mModel?.onUnload()

        clearAddOns()
        receiveEvents(false)
    }

    fun receiveEvents(receive: Boolean) {
        val eventManager: IEventManager? = getAddOn(AddOnType.EVENT_MANAGER) as IEventManager?

        if (receive) {
            mEventSubscriber = eventManager?.subscribe(callback = { event: Event ->
                onReceiveEvents(event)
            })
        } else {
            if (mEventSubscriber != null) {
                eventManager?.unsubscribe(mEventSubscriber)
            }
            mEventSubscriber = null
        }
    }

    open fun onReceiveEvents(event: Event) {
    }

    // Addons related methods.
    private val mAddOn: IAddOn by lazy {
        AddOn()
    }

    override fun getAddOn(type: String): IAddOn? {
        return mAddOn.getAddOn(type)
    }

    override fun getAddOns(): Map<String, IAddOn> {
        return mAddOn.getAddOns()
    }

    override fun addAddOn(type: String, addOn: IAddOn) {
        mAddOn.addAddOn(type, addOn)
    }

    override fun addAddOns(addons: Map<String, IAddOn>) {
        mAddOn.addAddOns(addons)
    }

    override fun removeAddOn(type: String) {
        mAddOn.removeAddOn(type)
    }

    override fun removeAddOns(types: List<String>) {
        mAddOn.removeAddOns(types)
    }

    override fun clearAddOns() {
        mAddOn.clearAddOns()
    }
}